### PR TITLE
[ML] Merge the pytorch-inference feature branch

### PR DESCRIPTION
**Pull request by davidkyle**
The feature branch contains changes to configure PyTorch models with a `TrainedModelConfig` and defines a format to store the binary models. The `_start` and `_stop` deployment actions control the model lifecycle and the model can be directly evaluated with the `_infer` endpoint. 2 Types of NLP tasks are supported: Named Entity Recognition and Fill Mask.

The feature branch consists of these PRs:
- #73523
- #72218
- #71679
- #71323
- #71035
- #71177
- #70713

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    "stability":"experimental",
+    "visibility":"public",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/json"]
+    },
+    "url":{
+      "paths":[
+        {
+          "path":"/_ml/trained_models/deployment/{model_id}/_start",
```
Comment (sethmlarson) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json:16]: Other paths for trained model APIs start with `/_ml/trained_models/{model_id}`, would it be more consistent to use `/_ml/trained_models/{model_id}/deployment/[_start|_stop]`?

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+  "ml.start_deployment":{
```
Comment (sethmlarson) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json:2]: Doesn't mention "trained model` in the API name, would have to be discovered via parameter names. Perhaps `ml.[start|stop]_trained_model_deployment`?

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.stop_deployment.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "ml.stop_deployment":{
+    "documentation":{
+      "url":"https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-stop-deployment.html",
+      "description":"Stpo a trained model deployment."
```
Comment (sethmlarson) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.stop_deployment.json:5]: typo

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Early review comments, very excited for this functionality.

**Comment by davidkyle** [comment]
Thanks for jumping in with an early review @sethmlarson

> use `/_ml/trained_models/{model_id}/deployment/[_start|_stop]`?

👍 This makes sense to me I'll raise it with the team


I've missed out the spec of the `_infer` endpoint which follows the same pattern as start & stop:  `/_ml/trained_models/deployment/{model_id}/_infer` and will add that soon. 

These APIs may be in flux for a short while as we work through all the use cases. Is that a problem for the clients team? Would you prefer us to tell you when we settled on something we like?

**Comment by sethmlarson** [comment]
@davidkyle It's no problem for us that these APIs may change especially if they're experimental/on `master` branch, we'll regenerate from the specs each time :) We like being a part of the process even if it's just seeing changes happen.

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json (hunk 1 above)
Comment (davidkyle) [reply rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json]: I switched the path to `/_ml/trained_models/{model_id}/deployment/_XXX` and added the _infer endpoint spec (ml.infer_trained_model_deployment.json)

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json (hunk 2 above)
Comment (davidkyle) [reply rest-api-spec/src/main/resources/rest-api-spec/api/ml.start_deployment.json]: Renamed ✔️

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.infer_trained_model_deployment.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "ml.infer_trained_model_deployment":{
+    "documentation":{
+      "url":"https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-infer-trained-model-deployment.html",
+      "description":"Evaluate a trained model."
```
Comment (sethmlarson) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.infer_trained_model_deployment.json:5]: Confirming this is the proper description for this API

**Review by sethmlarson — APPROVED** [review]
Looks good from an API spec perspective :tada: One comment I was unsure about.

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.infer_trained_model_deployment.json (hunk 4 above)
Comment (davidkyle) [reply rest-api-spec/src/main/resources/rest-api-spec/api/ml.infer_trained_model_deployment.json:5]: Yes 'evaluating' a model can be considered another way of saying 'performing inference' or even 'calling forward (propogation)'

Thanks for the review

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**